### PR TITLE
Initial modularization frame

### PR DESCRIPTION
**Pull request by dpc**
Right now it is not being used by any production code, but it prepares the general structure that will be used to port the existing code into.

### CODE REVIEW

File: core/api/src/encode.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+pub trait ModuleDecodable<M>: Sized {
+    /// Decode an object with a well-defined format
+    fn consensus_decode<R: std::io::Read>(
+        r: &mut R,
+        modules: &BTreeMap<ModuleKey, M>,
+    ) -> Result<Self, DecodeError>;
+}
+
+/*
+macro_rules! impl_module_decodable_forward_to_decodable {
```
Comment (dpc) [inline core/api/src/encode.rs:19]: Delete.

File: client/client-lib/Cargo.toml [hunk 2]
```diff
@@ -21,11 +21,14 @@ bitcoin = "0.28.1"
 bitcoin_hashes = "0.10.0"
 futures = "0.3.9"
 hex = "0.4.3"
+impl-tools = "0.4.0"
```
Comment (dpc) [inline client/client-lib/Cargo.toml:24]: We could probably do without it by just `impl Deref ...` in the macro.

File: core/api/src/transaction.rs [hunk 3]
```diff
@@ -0,0 +1 @@
+
```
Comment (dpc) [inline core/api/src/transaction.rs:1]: Delete this file.

File: core/api/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+/// Created by generating transaction [`Output`], spendable
+/// by converting to [`Input`].
+pub trait ModuleSpendableOutput: DynEncodable {
+    fn as_any(&self) -> &(dyn Any + '_);
+    /// Module key
+    fn module_key(&self) -> ModuleKey;
+    fn amount(&self) -> Amount;
+    fn clone(&self) -> SpendableOutput;
+
+    // TODO: move to be module function
```
Comment (dpc) [inline core/api/src/lib.rs:258]: Done. Delete.

File: core/api/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[lib]
+name = "fedimint_core_api"
+path = "src/lib.rs"
+
+[dependencies]
+async-trait = "0.1.57"
+bitcoin = "*"
```
Comment (dpc) [inline core/api/Cargo.toml:17]: Fill the `*`

**Review by elsirion — APPROVED** [review]
I think we should focus on the server side first (small changes=better), but if it helps to tackle the client at the same time this seems good enough. Didn't have time to dig too deep, but feel free to proceed.

File: fedimint-client/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                assets.push(
+                    module
+                        .decode_spendable_output(&mut raw_asset.as_slice())
+                        .expect("corrupted asset in the db"),
+                );
+            }
+        }
+
+        Ok(assets)
+    }
```
[the remark this answers is not in the document]
Comment (dpc) [reply fedimint-client/src/lib.rs:86]: This is just a sketch code for how loading these types from DB could work, just to make sure the types work etc. All for `example_workflow`. I might delete it. We're not going to be refactoring  client yet.

**Comment by dpc** [comment]
> I think we should focus on the server side first (small changes=better), but if it helps to tackle the client at the same time this seems good enough.

I only want the client crates ther so the complete frame is there for reference and to make sure high level details (types, generics, no cyclical depedencies, etc.) work for both sides. I will not be touching client code yet.

**Comment by dpc** [comment]
@elsirion I still intend to land. I removed the placeholder parts that you pointed out. Let me know what you think.

**Comment by dpc** [comment]
@jkitman You're opinion will be greatly appreciated. Basically I want to land all this to get it out of the way, and be able to gradually refactor the real code towards the end goal.

**Comment by dpc** [comment]
Landed as #741

**Pull request closed without merging**